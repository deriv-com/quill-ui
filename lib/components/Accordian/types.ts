import { ReactNode } from "react";
import { AccordionVariants } from ".";
import { TRegularSizes } from "@types";

export type AccordionProps = {
    id?: string;
    className?: string;
    title?: string;
    subtitle?: string;
    icon?: ReactNode;
    color?: string;
    textSize?: TRegularSizes;
    expanded?: boolean;
    divider?: AccordionDivider;
    disabled?: boolean;
    contentClassname?: string;
    expandedColor?: TExpandedColor;
    content?: () => ReactNode;
    customContent?: () => ReactNode;
    isFlush?: boolean;
    onExpand?: (e: boolean, i: string) => void;
};
type TExpandedColor = "white" | "gray";
export interface AccordionTabProps extends Omit<AccordionProps, "content"> {
    tab?: Tab;
    className?: string;
    variant?: keyof AccordionVariants;
    content: {
        className?: string;
        data: AccordionProps[][];
    };
    multiCollapse?: boolean;
}

export type Tab = {
    align?: TabAlign;
    data: TabData[];
};

export type TabData = {
    id: number;
    title: string;
};

export type TabAlign = "center" | "end";

export type AccordionDivider = "none" | "both" | "bottom";
