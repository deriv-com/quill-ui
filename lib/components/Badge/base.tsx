import { forwardRef } from "react";
import { BadgeProps } from "./types";
import clsx from "clsx";
import "./badge.scss";
import { CaptionText, Text } from "@components/Typography";

export const BadgeBase = forwardRef<HTMLSpanElement, BadgeProps>(
    (
        { size = "sm", color = "success", label, variant, className, ...rest },
        ref,
    ) => {
        const labelColor = "badge__base-color";
        return (
            <span
                className={clsx(
                    className,
                    "badge__base",
                    `badge__variant__${variant}--size-${size}`,
                    `badge__color--${color}`,
                )}
                ref={ref}
                {...rest}
            >
                {label &&
                    (size === "sm" ? (
                        <CaptionText bold color={labelColor}>
                            {label}
                        </CaptionText>
                    ) : size === "md" ? (
                        <Text bold color={labelColor}>
                            {label}
                        </Text>
                    ) : size === "lg" ? (
                        <Text size="xl" bold color={labelColor}>
                            {label}
                        </Text>
                    ) : null)}
            </span>
        );
    },
);
export default BadgeBase;
