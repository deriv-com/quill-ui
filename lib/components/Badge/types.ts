import { TRegularSizes } from "@types";

export type TBadgeColorVariants = "success" | "warning" | "danger";
export type TBadgeVariant = "status" | "notification";

export type BadgeProps = {
    variant?: TBadgeVariant;
    size?: TRegularSizes;
    color?: TBadgeColorVariants;
    className?: string;
    label?: string;
    badgeClassName?: string;
};
