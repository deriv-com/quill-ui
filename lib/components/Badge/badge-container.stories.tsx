import type { Meta, StoryObj } from "@storybook/react";
import BadgeContainer from "./index";
import { LabelPairedBellXlRegularIcon } from "@deriv/quill-icons/LabelPaired";
import { StandaloneWhatsappIcon } from "@deriv/quill-icons/Standalone";
import { TextField } from "..";

const meta = {
    title: "Components/Badge/Conatiner",
    component: BadgeContainer,
    parameters: {
        layout: "centered",
    },
    argTypes: {
        size: {
            options: ["sm", "md", "lg"],
            control: { type: "radio" },
            description: "To select the size of the badge",
        },
        color: {
            options: ["success", "warning", "danger"],
            control: { type: "radio" },
            description: "To select the color of the badge",
        },
    },
    tags: ["autodocs"],
} satisfies Meta<typeof BadgeContainer>;

export default meta;
type Story = StoryObj<typeof meta>;

export const NotificationOnTop: Story = {
    args: {
        children: <LabelPairedBellXlRegularIcon />,
        variant: "notification",
        label: "3",
        size: "sm",
        color: "danger",
        position: "top-right",
    },
};

export const StatusOnBottom: Story = {
    args: {
        variant: "status",
        position: "bottom-right",
        children: <StandaloneWhatsappIcon fill="#000000" iconSize="xl" />,
        size: "sm",
        color: "success",
    },
};
export const NotificationAtCenter: Story = {
    args: {
        children: <TextField placeholder="Notification"></TextField>,
        variant: "notification",
        position: "center",
        size: "md",
        label: "10",
        color: "danger",
    },
};
