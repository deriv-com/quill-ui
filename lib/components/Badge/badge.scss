.badge {
    &__container {
        position: relative;

        &-children {
            height: 100%;
            width: 100%;
        }
    }
    &__position {
        &-bottom {
            &-sm {
                bottom: -2px;
                right: -2px;
            }
            &-md {
                bottom: -4px;
                right: -4px;
            }
            &-lg {
                bottom: -8px;
                right: -8px;
            }
        }
        &-top {
            &-sm {
                top: -9px;
                right: -9px;
            }
            &-md {
                top: -9px;
                right: -9px;
            }
            &-lg {
                top: -12px;
                right: -12px;
            }
            &-xl {
                top: -12px;
                right: -12px;
            }
            &-2xl {
                top: -18px;
                right: -18px;
            }
           
        }
        &-center {
            top: var(--core-size-800);
            right: var(--core-size-700);
        }
        
    }

    &__base {
        display: inline-flex;
        position: absolute;
        border-radius: var(--core-borderRadius-4800);
        align-items: center;
        justify-content: center;
        color: var(--core-color-solid-slate-50);
    }

    &__variant {
        &__status {
            &--size {
                &-sm {
                    height: var(--core-size-400);
                    min-width: var(--core-size-400);
                }

                &-md {
                    height: var(--core-size-800);
                    min-width: var(--core-size-800);
                }

                &-lg {
                    height: var(--core-size-1200);
                    min-width: var(--core-size-1200);
                }
            }
        }

        &__notification {
            &--size {
                &-sm {
                    padding-inline: var(--temp-static-spacing-5);
                    height: var(--core-size-900);
                    min-width: var(--core-size-900);
                }

                &-md {
                    padding-inline: var(--temp-static-spacing-7);
                    height: var(--core-size-1200);
                    min-width: var(--core-size-1200);
                }

                &-lg {
                    padding-inline: var(--core-spacing-500);
                    height: var(--core-size-1800);
                    min-width: var(--core-size-1800);
                }
            }
        }
    }

    &__color {
        &--success {
            background-color: var(--core-color-solid-green-700);
        }

        &--warning {
            background-color: var(--core-color-solid-yellow-700);
        }

        &--danger {
            background-color: var(--core-color-solid-red-700);
        }
    }
}