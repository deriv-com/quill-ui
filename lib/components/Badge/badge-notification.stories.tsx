import type { Meta, StoryObj } from "@storybook/react";
import Badge from "./base";

const meta = {
    title: "Components/Badge/Notification",
    component: Badge,
    parameters: {
        layout: "centered",
    },
    argTypes: {
        size: {
            options: ["sm", "md", "lg"],
            control: { type: "radio" },
            description: "To select the size of the badge",
        },
        color: {
            options: ["success", "warning", "danger"],
            control: { type: "radio" },
            description: "To select the color of the badge",
        },
    },
    tags: ["autodocs"],
} satisfies Meta<typeof Badge>;

export default meta;
type Story = StoryObj<typeof meta>;

export const SuccessBadge: Story = {
    args: {
        variant: "notification",
        label: "1",
        size: "sm",
        color: "success",
    },
};
export const warningBadge: Story = {
    args: {
        variant: "notification",
        label: "2",
        size: "md",
        color: "warning",
    },
};
export const DangerBadge: Story = {
    args: {
        variant: "notification",
        label: "3",
        size: "lg",
        color: "danger",
    },
};
