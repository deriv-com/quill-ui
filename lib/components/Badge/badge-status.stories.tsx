import type { Meta, StoryObj } from "@storybook/react";
import Badge from "./index";

const meta = {
    title: "Components/Badge/Status",
    component: Badge,
    parameters: {
        layout: "centered",
    },
    argTypes: {
        size: {
            options: ["sm", "md", "lg"],
            control: { type: "radio" },
            description: "To select the size of the badge",
        },
        color: {
            options: ["success", "warning", "danger"],
            control: { type: "radio" },
            description: "To select the color of the badge",
        },
    },
    tags: ["autodocs"],
} satisfies Meta<typeof Badge>;

export default meta;
type Story = StoryObj<typeof meta>;

export const SuccessBadge: Story = {
    args: {
        variant: "status",
        size: "sm",
        color: "success",
    },
};
export const warningBadge: Story = {
    args: {
        variant: "status",
        size: "md",
        color: "warning",
    },
};
export const DangerBadge: Story = {
    args: {
        variant: "status",
        size: "lg",
        color: "danger",
    },
};
