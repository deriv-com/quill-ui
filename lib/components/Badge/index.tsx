import { ReactNode, forwardRef } from "react";
import BadgeBase from "./base";
import { BadgeProps } from "./types";
import "./badge.scss";
export interface TBadgeContainer extends BadgeProps {
    position?: "top-right" | "center" | "bottom-right";
    children?: ReactNode;
    badgeClassName?: string;
}

export const BadgeContainer = forwardRef<HTMLDivElement, TBadgeContainer>(
    (
        {
            size,
            color,
            label,
            children,
            variant,
            position,
            badgeClassName,
            ...rest
        },
        ref,
    ) => {
        const badge__position =
            position === "bottom-right"
                ? "badge__position-bottom"
                : position === "top-right"
                  ? "badge__position-top"
                  : "badge__position-center";
        return (
            <div className="badge__container">
                <div className="badge__container-children">
                    {children && children}
                </div>
                <BadgeBase
                    {...rest}
                    size={size}
                    color={color}
                    label={label}
                    variant={variant}
                    badgeClassName={badgeClassName}
                    className={badge__position}
                    ref={ref}
                ></BadgeBase>
            </div>
        );
    },
);

export default BadgeContainer;
