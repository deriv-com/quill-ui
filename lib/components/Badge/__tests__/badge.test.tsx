import { render, screen } from "@testing-library/react";
import Badge from "../base";
import { TRegularSizes } from "@types";
import { TBadgeColorVariants } from "../types";

describe("<Badge.Empty/>", () => {
    it("should render badge component correctly", () => {
        render(<Badge aria-label="Status Badge" variant="status" />);
        const badge = screen.getByLabelText("Status Badge");
        expect(badge).toBeInTheDocument();
    });

    it("should render variant correctly", () => {
        const statusVariant = "status";
        render(
            <Badge aria-label={`${statusVariant}`} variant={statusVariant} />,
        );
        const statusBadge = screen.getByLabelText(statusVariant);
        expect(statusBadge).toBeInTheDocument();
    });

    const sizes: TRegularSizes[] = ["sm", "md", "lg"];
    sizes.forEach((size) => {
        it(`should render badge correctly with size ${size}`, () => {
            render(
                <Badge
                    size={size}
                    aria-label={`${size} Badge`}
                    variant="status"
                />,
            );
            const badge = screen.getByLabelText(`${size} Badge`);
            expect(badge).toMatchSnapshot();
        });
    });

    const colors: TBadgeColorVariants[] = ["danger", "success", "warning"];
    colors.forEach((color) => {
        it(`should render badge correctly with color ${color}`, () => {
            render(
                <Badge
                    color={color}
                    aria-label={`${color} Badge`}
                    variant="notification"
                />,
            );
            const badge = screen.getByLabelText(`${color} Badge`);
            expect(badge).toMatchSnapshot();
        });
    });
});
