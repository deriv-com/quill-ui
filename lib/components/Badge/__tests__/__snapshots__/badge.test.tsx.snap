// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge.Empty/> should render badge correctly with color danger 1`] = `
<span
  aria-label="danger Badge"
  class="badge__base badge__variant__notification--size-sm badge__color--danger"
/>
`;

exports[`<Badge.Empty/> should render badge correctly with color success 1`] = `
<span
  aria-label="success Badge"
  class="badge__base badge__variant__notification--size-sm badge__color--success"
/>
`;

exports[`<Badge.Empty/> should render badge correctly with color warning 1`] = `
<span
  aria-label="warning Badge"
  class="badge__base badge__variant__notification--size-sm badge__color--warning"
/>
`;

exports[`<Badge.Empty/> should render badge correctly with size lg 1`] = `
<span
  aria-label="lg Badge"
  class="badge__base badge__variant__status--size-lg badge__color--success"
/>
`;

exports[`<Badge.Empty/> should render badge correctly with size md 1`] = `
<span
  aria-label="md Badge"
  class="badge__base badge__variant__status--size-md badge__color--success"
/>
`;

exports[`<Badge.Empty/> should render badge correctly with size sm 1`] = `
<span
  aria-label="sm Badge"
  class="badge__base badge__variant__status--size-sm badge__color--success"
/>
`;
